rustworkx-core/src/dag_algo.rs

```diff
@@ -314,6 +319,176 @@ where
     Ok(Some((path, path_weight)))
 }
 
+/// Return a list of graph layers
+///
+/// A layer is a subgraph whose nodes are disjoint, i.e.,
+/// a layer has depth 1. The layers are constructed using a greedy algorithm.
+///
+/// Arguments:
+///
+/// * `graph` - The graph to get the layers from
+/// * `first_layer` - A list of node ids for the first layer. This
+///     will be the first layer in the output
+///
+/// ```
+/// use rustworkx_core::petgraph::prelude::*;
+/// use rustworkx_core::dag_algo::layers;
+/// use rustworkx_core::dictmap::*;
+///
+/// let edge_list = vec![
+///  (0, 1),
+///  (1, 2),
+///  (2, 3),
+///  (3, 4),
+/// ];
+///
+/// let graph = DiGraph::<u32, u32>::from_edges(&edge_list);
+/// let layers: Vec<Vec<NodeIndex>> = layers(&graph, vec![0.into(),1.into()]).unwrap().collect();
+/// let expected_layers: Vec<Vec<NodeIndex>> = vec![
+///     vec![0.into(),1.into()],
+///     vec![1.into(),2.into()],
+///     vec![2.into(),3.into()],
+///     vec![3.into(),4.into()],
+///     vec![4.into()]
+/// ];
+/// assert_eq!(layers, expected_layers)
+/// ```
+pub fn layers<G>(
+    graph: G,
+    first_layer: Vec<G::NodeId>,
+) -> Result<LayersIter<G, G::NodeId>, LayersError>
```

Comment:
If possible I would do this as

```suggestion
) -> Result<impl Iterator<Item=Vec<G, G::NodeId>>, LayersError>
```

It leaves us the flexibility to leave `LayersIter` private and also tweak the implementation in the future if needed. We basically are making the guarantee that the API will just return an iterator of vecs of node ids.